IPIP/0000-double-hash-dht.md

```diff
@@ -0,0 +1,337 @@
+
+Ephemeral `PeerID`s references:
+- https://github.com/libp2p/libp2p/issues/37
+
+The other alternative to increase the Reader Privacy level in the IPFS DHT is the use of Mixnets such as Tor or I2P. Mixnets usually provide an excellent Reader- and Writer Privacy level, but lookup latency is significantly higher. Hence the use of Mixnets is generally not good for all use cases, but only when strong privacy guarantees are required. Mixnets can easily be built on top of the Double Hash DHT to maximize user Privacy. IPFS users willing to remain pseudonymous could use the existing Tor network to hide their identity. Another alternative could be to create a Mixnet out of the IPFS network, e.g include mixing capabilities in every libp2p host. There has been some ongoing work on IPFS-Tor integration.
+
+Mixnets references:
+- Berty's [go-libp2p-tor-transport](https://github.com/berty/go-libp2p-tor-transport)
+- [Hosting an IPFS Gateway Through a Tor Proxy](https://www.minds.com/raymondsmith98/blog/tutorial-tor-hosting-an-ipfs-gateway-through-a-tor-proxy-857369540936916992)
+- Mixnet and Content Routing ([IPFS Thing 2022 Video](https://www.youtube.com/watch?v=f85U8b5g-Ks), [Notes](https://hackmd.io/@nZ-twauPRISEa6G9zg3XRw/BkrcMOLd9)) by [noot](https://github.com/noot)
+- [Nym Mixnet](https://nymtech.net/)
+- https://github.com/ipfs/notes/issues/37
+
+
+## Open Questions
+
+- If we plan to move to using SHA3 instead of SHA2 to generate 256-bits digests, this migration is the perfect opportunity, as we will be breaking everything anyways. SHA3 was proved to be more secure against Length Extension Attacks. It has not be proven whether SHA2 or SHA3 is more collision resistant and secure against preimage attacks. See this [comparison](https://en.wikipedia.org/wiki/SHA-3#Comparison_of_SHA_functions).
+- Is it wise to encrypt the `CPPeerID` using `MH` directly? It would be possible to derive another identifier from `MH` (such as `Hash("SOME_CONSTANT" || MH)`). `MH` is the master identifier of the content, hence if it is revealed all other identifiers can trivially be found. However, it is computationally impossible to recover `MH` from `Hash("SOME_CONSTANT" || MH)`.
+- It may be fine to use `TS` as nonce, it spares bytes on the wire. However, if two Content Providers publish the same content at the same time (`TS` either in seconds or milliseconds), then the DHT Server may be able to forge a valid Provider Records for itself.
+- As multiple `HASH2` match each `Prefix` and the Client is only interested in a single one, should we send the `HASH2` along with each encrypted provider record (network load overhead) or let the Client try to decrypt all payloads and see for themselves which one opens (cpu overhead)?
```

Comment:
As multiple `HASH2` match each `Prefix` and the Client is only interested in a single one, should we send the `HASH2` along with each encrypted provider record (network load overhead) or let the Client try to decrypt all payloads and see for themselves which one opens (cpu overhead)?

An alternative could be to index the encrypted provider records by some short suffix of `HASH2` (small network overload and less decryption to be performed on client side).

IPNI faces the same challenge for prefix requests. Any thoughts? @masih @ischasny @willscott

Reply:
> Should we send the HASH2 along with each encrypted provider record 

the IPNI non-streaming APIs actually group results by multihash. The rationale is to support batch lookup of multihashes.

Having said that I do think this needs more thought considering other works in the making like streaming support for HTTP delegated routing. In the interest of reducing scope, do you think it would be reasonable to separate that into its own IPIP? Would it be fair to say that prefix-based lookup is not a prerequisite for double-hashing reader privacy preservation?

Reply:
Even when grouped by multihash (`HASH2`) if a Prefix Request matches `k` provider records, adding the multihash in the payload would add 32*k bytes on the wire, that aren't strictly necessary.

In a sense the `HASH2` is already contained in the encrypted payload. If you are able to decrypt it (using `MH`) then you know to which `CID` or `HASH2` the content is associated.

Prefix requests add non negligible privacy guarantees (k-anonymity, plausible deniability), so I think that breaking the Spec in 2 IPIP wouldn't be relevant here. We could open a specific IPIP for Prefix Requests, and then report back any decision in this IPIP.

Reply:
See https://github.com/guillaumemichel/specs/blob/79a915fe252dd8ddd5a8c6d923dc00a8c47ccd0a/IPIP/0373-double-hash-dht.md?plain=1#L89